nexus/src/sagas.rs

```diff
@@ -308,3 +321,196 @@ async fn sic_instance_ensure(
+    type ExecContextType = Arc<SagaContext>;
+}
+
+pub fn saga_instance_migrate() -> SagaTemplate<SagaInstanceMigrate> {
+    let mut template_builder = SagaTemplateBuilder::new();
+
+    template_builder.append(
+        "migrate_id",
+        "GenerateMigrateId",
+        new_action_noop_undo(saga_generate_uuid),
+    );
+
+    template_builder.append(
+        "migrate_prep",
+        "MigratePrep",
+        new_action_noop_undo(sim_migrate_prep),
+    );
+
+    template_builder.append(
+        "dst_propolis_id",
+        "GeneratePropolisId",
+        new_action_noop_undo(saga_generate_uuid),
+    );
+
+    template_builder.append(
+        "instance_migrate",
+        "InstanceMigrate",
+        // TODO robustness: This needs an undo action
+        new_action_noop_undo(sim_instance_migrate),
+    );
+
+    template_builder.append(
+        "cleanup_source",
+        "CleanupSource",
+        // TODO robustness: This needs an undo action. Is it even possible
+        // to undo at this point?
+        new_action_noop_undo(sim_cleanup_source),
+    );
+
+    template_builder.build()
+}
+
+async fn sim_migrate_prep(
+    sagactx: ActionContext<SagaInstanceMigrate>,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    let migrate_uuid = sagactx.lookup::<Uuid>("migrate_id")?;
+
+    let instance = osagactx
+        .datastore()
+        .instance_fetch_by_name(
+            &params.project_id,
+            &params.instance_name.clone().into(),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let instance_id = instance.id();
+
+    let runtime: InstanceRuntimeState = instance.runtime_state.into();
+
+    // Is an existing migration ongoing?
+    if runtime.run_state == InstanceState::Migrating {
+        match runtime.migration_uuid {
+            // This is the same migration, nothing else to do
+            Some(id) if id == migrate_uuid => {
+                return Ok(());
+            }
+            // Encountered a different ongoing migration, bail
+            Some(_) => {
+                return Err(ActionError::action_failed(
+                    "unexpected migration already in progress".to_string(),
+                ));
+            }
+            // Marked as migrating but no ID yet?
+            // Treat as new migration
+            None => {}
+        }
+    }
+
+    // Update the Instance's runtime state to indicate it's currently migrating.
+    // This also acts as a lock somewhat to prevent any further state changes
+    // from being performed on the instance in the meantime.
+    // See `check_runtime_change_allowed`.
+    let runtime = InstanceRuntimeState {
+        run_state: InstanceState::Migrating,
+        migration_uuid: Some(migrate_uuid),
+        ..runtime
+    };
+    osagactx
+        .datastore()
+        .instance_update_runtime(&instance_id, &runtime.into())
```

Comment:
Would failing the action if the update failed (`false` returned) here be reasonable?

Reply:
Yeah, I think failing the action would be necessary! If we can't mark the instance as migrating, that opens up a can-of-worms of unsafety to keep pushing forward anyway.

(Seems like that's the current behavior, right? Mapping the error to `action_failed`?)

Reply:
Ah, I meant specifically acting on the bool return value from `instance_update_runtime` indicating Updated (true) or NotUpdatedButExists (false). I'm inclined to fail if we get false, for more or less the same reason you outlined.

Reply:
Ah, right, that would happen if the [generation number](https://rfd.shared.oxide.computer/rfd/0192#_transactions_ctes_sagas_generation_numbers) was modified between originally reading the instance, like if some other concurrent operation (or a state change notification from the sled agent) occurred on this object.

I think we have two options:

1. Fail the operation, claiming that a concurrent operation happened, and we weren't able to trigger the update. Note that the other uses of this function don't typically need to do this, because they can treat the earlier concurrent operation as the "winner".

2. Use a different query operation to the datastore that avoids the situation altogether. `instance_update_runtime` does not bother checking the additional constraints which you use - namely, are we migrating, and with the right migration ID? - so performing this "read - check - modify" operation out of a single query introduces the possibility of the generation number being bumped. 

If we created a variant of `instance_update_runtime` which also `.filter`ed on `dsl::state.ne(migrating)`, and `dsl::migration_id.is_null()`, we could perform the read and write within the same operation, avoiding the possibility of concurrency. (Since we are always able to read the old value, we could handle the idempotency cases, like "what if we already started the migration with this exact UUID", *after* querying the DB).

I kinda think the second option would be ideal?

Reply:
Admittedly, there's also a third option - "use a transaction" - which is a bit more heavyweight, and which would be reasonable if we had to check constraints that don't exist in the DB. However, in this case, I believe all the state you're checking is in the same table, so this seems similar-but-worse-than option (2).

Reply:
I also filed https://github.com/oxidecomputer/omicron/issues/611 with some detail. I think this problem is exacerbated by your (100% reasonable) usage here, which sets `migration_id` in addition to updating the state field.

Reply:
Hrm, sounds like the takeaway from that issue is actually a totally different option: Tell the Sled Agent to try to start the Migration, and only update the DB once it confirms that it has done so.

Reply:
Yup, that makes sense in keeping with Sled Agent as the source of truth for instance state.